### PR TITLE
Add code coverage through codecov.io

### PR DESCRIPTION
**Pull request by ohm314**
**Description**

This PR introduces code coverage monitoring into the CI. Reports are provided by codecov.io

**How to test this?**

Check github actions on `ci_test` branch. You can also run locally the commands in `.github/workflows/ci.yml`.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -122,3 +122,33 @@ jobs:
       working-directory: ${{github.workspace}}/build
       shell: bash -l {0}
       run: ctest --output-on-failure -C $BUILD_TYPE
+
+  # Plan for preoducing code coverage report
```
Comment (matz-e) [inline .github/workflows/ci.yml:126]:
```suggestion
  # Job producing code coverage report
```

Comment (ohm314) [reply .github/workflows/ci.yml:126]: that's what I get fro copy-pasting... :P

**Review by matz-e — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -122,3 +122,33 @@ jobs:
       working-directory: ${{github.workspace}}/build
       shell: bash -l {0}
       run: ctest --output-on-failure -C $BUILD_TYPE
+
+  # Job preoducing code coverage report
+  # =========================================
```
Comment (ferdonline) [inline .github/workflows/ci.yml:127]:
```suggestion
  # Job producing code coverage report
  # ==================================
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -122,3 +122,33 @@ jobs:
...
+    - name: Build and test for code coverage
+      run: |
+        cmake -E make_directory $GITHUB_WORKSPACE/build
+        cd $GITHUB_WORKSPACE/build
+        cmake $GITHUB_WORKSPACE \
+          -DCMAKE_BUILD_TYPE=$BUILD_TYPE \
+          -DCMAKE_INSTALL_PREFIX=$INSTALL_DIR \
+          -DHIGHFIVE_USE_EIGEN:BOOL=TRUE \
+          -DHIGHFIVE_BUILD_DOCS:BOOL=FALSE \
+          -DCMAKE_CXX_FLAGS="-coverage -O0"
```
Comment (ferdonline) [inline .github/workflows/ci.yml:148]: Instead of repeating all options, use the `CMAKE_OPTIONS` variable (see above)

**Review by ferdonline — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 3 above)
Comment (ohm314) [reply .github/workflows/ci.yml]: the whole thing with quoted arguments is just a mess, I hope the way I solved is ok.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -41,12 +41,10 @@ jobs:
...
     - name: Build
-      env:
-        CMAKE_OPTIONS: "\
-          -DHIGHFIVE_PARALLEL_HDF5:BOOL=ON \
-          -DHIGHFIVE_USE_XTENSOR:BOOL=ON"
       shell: bash -l {0}  # required to activate conda
-      run: source $GITHUB_WORKSPACE/.github/build.sh
+      run: |
+        CMAKE_OPTIONS: (-DHIGHFIVE_PARALLEL_HDF5:BOOL=ON -DHIGHFIVE_USE_XTENSOR:BOOL=ON)
+        source $GITHUB_WORKSPACE/.github/build.sh
```
Comment (matz-e) [inline .github/workflows/ci.yml:47]: Now that this pops up… any particular reason for `source`? I've seen quite a few issues arise when `source` is used to execute other scripts.

Comment (ferdonline) [reply .github/workflows/ci.yml:47]: build is a small script, it's convenient to have it running with the full context, so literally we just move the code to a file on the side

**Review by ferdonline — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -115,10 +113,33 @@ jobs:
...
-        CMAKE_OPTIONS="-DHIGHFIVE_USE_XTENSOR:BOOL=ON"
+        CMAKE_OPTIONS=(-DHIGHFIVE_USE_XTENSOR:BOOL=ON)
         source $GITHUB_WORKSPACE/.github/build.sh
 
     - name: Test
       working-directory: ${{github.workspace}}/build
       shell: bash -l {0}
       run: ctest --output-on-failure -C $BUILD_TYPE
+
+  # Job producing code coverage report
```
Comment (alexsavulescu) [inline .github/workflows/ci.yml:124]: I'd move the job in its own file given GHA experience. Fine with me either way

**Review by alexsavulescu — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 4 above)
Comment (ohm314) [reply .github/workflows/ci.yml:47]: Yeah, I was hesitant about `source` but I agree with Fernando that in this context, the build script is so trivial, that this is just fine

File: .github/workflows/ci.yml (hunk 5 above)
Comment (ohm314) [reply .github/workflows/ci.yml:124]: we can do this at some point when the pipeline grows to clean up a bit. I'm gonna merge as is.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (matz-e) [reply .github/workflows/ci.yml:47]: OK, just making sure :)

**Pull request merged**